### PR TITLE
Processing data

### PR DESCRIPTION
**Pull request by Daphineatim**
### In this milestone, the following were implemented;

- Created model classes for all entities as shown in the [ERD diagram](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/images/blog_app_erd.png).
- Set up associations between models.

- [ ] Created custom methods

- Users
A method that returns the 3 most recent posts for a given user.
Posts
A method that updates the posts counter for a user.
A method which returns the 5 most recent comments for a given post.
- Comments
A method that updates the comments counter for a post.
- Likes
A method that updates the likes counter for a post.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -1,2 +1,9 @@
 class User < ApplicationRecord
+  has_many :posts
+  has_many :likes
+  has_many :comments
```
Comment (Deepakdanger) [inline app/models/user.rb]: - [x] All model has author_id then kindly use :-
```
has_many :posts, foreign_key: 'author_id'
has_many :comments, foreign_key: 'author_id'
has_many :likes, foreign_key: 'author_id'
```

File: app/models/like.rb [hunk 2]
```diff
@@ -1,2 +1,10 @@
 class Like < ApplicationRecord
+  belongs_to :user
```
Comment (Deepakdanger) [inline app/models/like.rb]: - [x] Kindly use `belongs_to :author, class_name: 'User'` as author_id is present in `Like`.

File: app/models/user.rb [hunk 3]
```diff
@@ -1,2 +1,9 @@
...
+
+  def last_3_posts
+    Post.where(user_id: id).order(created_at: :desc).limit(3)
```
Comment (Deepakdanger) [inline app/models/user.rb]: - [x] There is association b/w user & posts so can directly use
`posts.order(created_at: :desc).limit(3)`

File: app/models/post.rb [hunk 4]
```diff
@@ -1,2 +1,15 @@
...
+  has_many :likes
+
+  def last_5_comments
+    Comment.where(post_id: id).order(created_at: :desc).limit(5)
+  end
+
+  def update_post_count_for_user
+    user = User.find_by(id: user_id)
+    counter = Post.where(user_id: user.id).count
+    user.update(PostsCounter: counter)
```
Comment (Deepakdanger) [inline app/models/post.rb]: - [x] There is an association between the post and the user, we can use direct author & increment by 1
`author.increment!(:posts_counter)`

File: app/models/like.rb [hunk 5]
```diff
@@ -1,2 +1,10 @@
...
+  belongs_to :post
+
+  def update_like_counter_for_post
+    post = Post.find_by(id: post_id)
+    count = Like.where(post_id: post.id).count
+    post.update(LikesCounter: count)
```
Comment (Deepakdanger) [inline app/models/like.rb]: - [x] There is an association between the post and the like, we can use direct posts & increment by 1
`post.increment!(:likes_counter)`

File: app/models/comment.rb [hunk 6]
```diff
@@ -1,2 +1,10 @@
 class Comment < ApplicationRecord
+  belongs_to :author, class_name: 'User'
+  belongs_to :post
+
+  def update_comment_counter_for_post
+    post = Post.find_by(id: post_id)
+    count = Comment.where(post_id: post.id).count
+    post.update(CommentsCounter: count)
```
Comment (Deepakdanger) [inline app/models/comment.rb]: - [x] There is an association between the post and the comment, we can use direct posts & increment by 1
`post.increment!(:`comments_counter`)`

**Review by Deepakdanger — CHANGES_REQUESTED** [review]
Hi @Daphineatim , 

Good job so far! 👏
There are some issues that you still need to work on to go to the next project but you are almost there!

## To Highlight :100: 
- No linter error :heavy_check_mark: 
- Website looks professional :+1: 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/user.rb (hunk 1 above)
Comment (Daphineatim) [reply app/models/user.rb]: Thank you so much, I have made the changes.

File: app/models/user.rb (hunk 3 above)
Comment (Daphineatim) [reply app/models/user.rb]: Thank you so much, I have made the changes.

File: app/models/post.rb (hunk 4 above)
Comment (Daphineatim) [reply app/models/post.rb]: Thank you so much.

File: app/models/like.rb (hunk 5 above)
Comment (Daphineatim) [reply app/models/like.rb]: Issue resolved, Thank you.

File: app/models/comment.rb (hunk 6 above)
Comment (Daphineatim) [reply app/models/comment.rb]: Thank you so much, Issue resolved.

**Review by codepantha — APPROVED** [review]
# STATUS: APPROVED 🟢 ✔️ 

## Hi @Daphineatim 👋, 

Well done for implementing the required changes 👏 
Your project is now complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**